// ng-chat/esm2015/ng-chat/components/ng-chat-options/ng-chat-options.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { Window } from '../../core/window';
export class NgChatOptionsComponent {
    constructor() {
        this.activeOptionTrackerChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} option
     * @return {?}
     */
    onOptionClicked(option) {
        if (option.action) {
            option.isActive = true;
            option.action(this.chattingTo);
            this.activeOptionTrackerChange.emit(option);
        }
    }
}
NgChatOptionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-chat-options',
                template: "<div *ngIf=\"options && options.length > 0\" class=\"ng-chat-options\">\r\n  <button class=\"ng-chat-options-activator\">\r\n    <span class=\"primary-text\">...</span>\r\n  </button>\r\n  <div class=\"ng-chat-options-content primary-background shadowed\">\r\n    <a *ngFor=\"let option of options; let i = index\" [ngClass]=\"'primary-text'\" (click)=\"onOptionClicked(option)\">\r\n      {{option.displayLabel}}\r\n    </a>\r\n  </div>      \r\n</div>\r\n",
                styles: [".ng-chat-options-activator{background-color:unset;color:#fff;line-height:28px;border:none;position:relative}.ng-chat-options-activator>span{position:relative;top:-5px;left:0}.ng-chat-options{position:relative;display:inline-block}.ng-chat-options:hover .ng-chat-options-content{display:block}.ng-chat-options:hover .ng-chat-options-activator{background-color:#ddd}.ng-chat-options-content{display:none;position:absolute;min-width:160px;z-index:1}.ng-chat-options-content a:hover{background-color:#ddd}.ng-chat-options-content a{padding:6px 16px;text-decoration:none;display:block}@media only screen and (max-width:581px){.ng-chat-options-content{right:0}}"]
            }] }
];
/** @nocollapse */
NgChatOptionsComponent.ctorParameters = () => [];
NgChatOptionsComponent.propDecorators = {
    options: [{ type: Input }],
    activeOptionTracker: [{ type: Input }],
    activeOptionTrackerChange: [{ type: Output }],
    chattingTo: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgChatOptionsComponent.prototype.options;
    /** @type {?} */
    NgChatOptionsComponent.prototype.activeOptionTracker;
    /** @type {?} */
    NgChatOptionsComponent.prototype.activeOptionTrackerChange;
    /** @type {?} */
    NgChatOptionsComponent.prototype.chattingTo;
}
