// angular-jitsi-meet/lib/jitsi/app.js
// Generated by CoffeeScript 1.10.0
(function() {
  var APP, initialized;

  initialized = false;

  APP = {
    initAppObject: function() {
      this.UI = require('./UI');
      this.connectionquality = require('jitsi-meet/modules/connectionquality/connectionquality');
      this.statistics = require('jitsi-meet/modules/statistics/statistics');
      this.RTC = require('jitsi-meet/modules/RTC/RTC');
      this.desktopsharing = require('jitsi-meet/modules/desktopsharing/desktopsharing');
      this.xmpp = require('jitsi-meet/modules/xmpp/xmpp');
      this.settings = require('jitsi-meet/modules/settings/Settings');
      return initialized = true;
    },
    startJitsiServices: function(startDesktopSharing) {
      startDesktopSharing = startDesktopSharing || false;
      if (!initialized) {
        this.initAppObject();
      }
      if (startDesktopSharing) {
        this.desktopsharing.init();
      }
      this.RTC.start();
      this.xmpp.start();
      this.statistics.start();
      return this.connectionquality.init();
    },
    stopJitsiServices: function() {
      console.log('global jitsi shutdown');
      this.connectionquality.stopSendingStats();
      this.statistics.stop();
      this.desktopsharing.destroy();
      return this.RTC.stop();
    }
  };

  module.exports = APP;

}).call(this);
