// angular-jitsi-meet/lib/jitsi/UI.js
// Generated by CoffeeScript 1.10.0
(function() {
  var EventEmitter, UI, UIEvents, eventEmitter, roomName,
    slice = [].slice;

  EventEmitter = require('events');

  UIEvents = require('jitsi-meet/service/UI/UIEvents');

  eventEmitter = new EventEmitter;

  roomName = '';

  UI = {
    addListener: function(type, listener) {
      return eventEmitter.on(type, listener);
    },
    removeListener: function(type, listener) {
      return eventEmitter.removeListener(type, listener);
    },
    emit: function() {
      var params, type;
      type = arguments[0], params = 2 <= arguments.length ? slice.call(arguments, 1) : [];
      return eventEmitter.emit.apply(eventEmitter, [type].concat(slice.call(params)));
    },
    onMainVideoUpdated: function(resourceJid) {
      return eventEmitter.emit(UIEvents.SELECTED_ENDPOINT, resourceJid);
    },
    onVideoPinned: function(resourceJid) {
      return eventEmitter.emit(UIEvents.PINNED_ENDPOINT, resourceJid);
    },
    onVideoUnpinned: function() {
      return eventEmitter.emit(UIEvents.PINNED_ENDPOINT);
    },
    getRoomNode: function() {
      return roomName;
    },
    setRoomNode: function(name) {
      return roomName = name;
    }
  };

  module.exports = UI;

}).call(this);
