// ngx-validators/esm5/components/equal-to/equal-to.directive.js
import * as tslib_1 from "tslib";
import { Directive, forwardRef, Input } from "@angular/core";
import { NG_VALIDATORS } from "@angular/forms";
import { delay } from "rxjs/operators";
var EqualToDirective = /** @class */ (function () {
    function EqualToDirective() {
    }
    EqualToDirective_1 = EqualToDirective;
    EqualToDirective.prototype.validate = function (c) {
        var otherControl = typeof this.equalTo === "string" ? c.parent.get(this.equalTo) : this.equalTo;
        if (!this.subscription) {
            this.subscription = otherControl.valueChanges.pipe(delay(1)).subscribe(function () {
                c.updateValueAndValidity();
            });
        }
        return c.value !== otherControl.value ? { notEqualTo: true } : null;
    };
    EqualToDirective.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    EqualToDirective.prototype.ngOnChanges = function (changes) {
        if (changes.equalTo && !changes.equalTo.isFirstChange()) {
            this.onChange();
        }
    };
    EqualToDirective.prototype.registerOnValidatorChange = function (fn) {
        this.onChange = fn;
    };
    var EqualToDirective_1;
    tslib_1.__decorate([
        Input()
    ], EqualToDirective.prototype, "equalTo", void 0);
    EqualToDirective = EqualToDirective_1 = tslib_1.__decorate([
        Directive({
            selector: "[equalTo][ngModel], [equalTo][formControlName], [equalTo][formControl]",
            providers: [
                {
                    provide: NG_VALIDATORS,
                    // tslint:disable-next-line:no-forward-ref
                    useExisting: forwardRef(function () { return EqualToDirective_1; }),
                    multi: true,
                },
            ],
        })
    ], EqualToDirective);
    return EqualToDirective;
}());
export { EqualToDirective };
