// ngx-validators/esm2015/components/email/email.directive.js
var EmailValidatorDirective_1, EmailSuggestValidatorDirective_1;
import * as tslib_1 from "tslib";
import { Directive, Input, forwardRef } from "@angular/core";
import { NG_VALIDATORS } from "@angular/forms";
import { EmailValidators } from "./email-validators";
let EmailValidatorDirective = EmailValidatorDirective_1 = class EmailValidatorDirective {
    constructor() {
        this.email = "normal";
    }
    ngOnInit() {
        this.setValidator(this.email);
    }
    setValidator(type) {
        switch (type) {
            case "simple":
                this.validator = EmailValidators.simple;
                break;
            case "normal":
                this.validator = EmailValidators.normal;
                break;
            default:
                this.validator = EmailValidators.normal;
                break;
        }
    }
    ngOnChanges(changes) {
        if (changes.email && !changes.email.isFirstChange()) {
            this.setValidator(changes.email.currentValue);
            this.onChange();
        }
    }
    registerOnValidatorChange(fn) {
        this.onChange = fn;
    }
    validate(c) {
        return this.validator(c);
    }
};
tslib_1.__decorate([
    Input()
], EmailValidatorDirective.prototype, "email", void 0);
EmailValidatorDirective = EmailValidatorDirective_1 = tslib_1.__decorate([
    Directive({
        selector: "[email][formControlName],[email][formControl],[email][ngModel]",
        providers: [
            {
                provide: NG_VALIDATORS,
                // tslint:disable-next-line:no-forward-ref
                useExisting: forwardRef(() => EmailValidatorDirective_1),
                multi: true,
            },
        ],
    })
], EmailValidatorDirective);
export { EmailValidatorDirective };
let EmailSuggestValidatorDirective = EmailSuggestValidatorDirective_1 = class EmailSuggestValidatorDirective {
    ngOnInit() {
        this.validator = EmailValidators.suggest(this.emailSuggest);
    }
    ngOnChanges(changes) {
        if (changes.emailSuggest && !changes.emailSuggest.isFirstChange()) {
            this.validator = EmailValidators.suggest(changes.emailSuggest.currentValue);
            this.onChange();
        }
    }
    registerOnValidatorChange(fn) {
        this.onChange = fn;
    }
    validate(c) {
        return this.validator(c);
    }
};
tslib_1.__decorate([
    Input()
], EmailSuggestValidatorDirective.prototype, "emailSuggest", void 0);
EmailSuggestValidatorDirective = EmailSuggestValidatorDirective_1 = tslib_1.__decorate([
    Directive({
        selector: "[emailSuggest][formControlName],[emailSuggest][formControl],[emailSuggest][ngModel]",
        providers: [
            {
                provide: NG_VALIDATORS,
                // tslint:disable-next-line:no-forward-ref
                useExisting: forwardRef(() => EmailSuggestValidatorDirective_1),
                multi: true,
            },
        ],
    })
], EmailSuggestValidatorDirective);
export { EmailSuggestValidatorDirective };
