// ngx-validators/esm5/components/creditcard/creditcard.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, forwardRef } from "@angular/core";
import { NG_VALIDATORS } from "@angular/forms";
import { CreditCardValidators } from "./creditcard-validators";
var CreditCardValidatorDirective = /** @class */ (function () {
    function CreditCardValidatorDirective() {
        this.creditCard = "all";
    }
    CreditCardValidatorDirective_1 = CreditCardValidatorDirective;
    CreditCardValidatorDirective.prototype.ngOnInit = function () {
        this.setCreditcardValidator(this.creditCard);
    };
    CreditCardValidatorDirective.prototype.setCreditcardValidator = function (type) {
        switch (type) {
            case "all":
                this.validator = CreditCardValidators.isCreditCard;
                break;
            case "americanExpress":
                this.validator = CreditCardValidators.americanExpress;
                break;
            case "dinersclub":
                this.validator = CreditCardValidators.dinersclub;
                break;
            case "discover":
                this.validator = CreditCardValidators.discover;
                break;
            case "jcb":
                this.validator = CreditCardValidators.jcb;
                break;
            case "maestro":
                this.validator = CreditCardValidators.maestro;
                break;
            case "mastercard":
                this.validator = CreditCardValidators.mastercard;
                break;
            case "visa":
                this.validator = CreditCardValidators.visa;
                break;
            default:
                this.validator = CreditCardValidators.isCreditCard;
                break;
        }
    };
    CreditCardValidatorDirective.prototype.ngOnChanges = function (changes) {
        if (changes.creditCard && !changes.creditCard.isFirstChange()) {
            this.setCreditcardValidator(changes.creditCard.currentValue);
            this.onChange();
        }
    };
    CreditCardValidatorDirective.prototype.validate = function (c) {
        return this.validator(c);
    };
    CreditCardValidatorDirective.prototype.registerOnValidatorChange = function (fn) {
        this.onChange = fn;
    };
    var CreditCardValidatorDirective_1;
    tslib_1.__decorate([
        Input()
    ], CreditCardValidatorDirective.prototype, "creditCard", void 0);
    CreditCardValidatorDirective = CreditCardValidatorDirective_1 = tslib_1.__decorate([
        Directive({
            selector: "[creditCard][formControlName],[creditCard][formControl],[creditCard][ngModel]",
            providers: [
                {
                    provide: NG_VALIDATORS,
                    // eslint-disable-next-line @typescript-eslint/no-use-before-define
                    useExisting: forwardRef(function () { return CreditCardValidatorDirective_1; }),
                    multi: true,
                },
            ],
        })
    ], CreditCardValidatorDirective);
    return CreditCardValidatorDirective;
}());
export { CreditCardValidatorDirective };
