// jitsi-meet-angular/fesm2015/jitsi-meet-angular.js
import { __decorate } from 'tslib';
import { ɵɵdefineInjectable, Injectable, Component, NgModule } from '@angular/core';

let JitsiMeetAngularService = class JitsiMeetAngularService {
    constructor() { }
};
JitsiMeetAngularService.ɵprov = ɵɵdefineInjectable({ factory: function JitsiMeetAngularService_Factory() { return new JitsiMeetAngularService(); }, token: JitsiMeetAngularService, providedIn: "root" });
JitsiMeetAngularService = __decorate([
    Injectable({
        providedIn: 'root'
    })
], JitsiMeetAngularService);

let JitsiMeetAngularComponent = class JitsiMeetAngularComponent {
    constructor() { }
    ngOnInit() {
    }
};
JitsiMeetAngularComponent = __decorate([
    Component({
        selector: 'lib-jitsi-meet-angular',
        template: `
    <p>
      hello from jitsi-meet-angular!
    </p>
  `
    })
], JitsiMeetAngularComponent);

let JitsiMeetAngularModule = class JitsiMeetAngularModule {
};
JitsiMeetAngularModule = __decorate([
    NgModule({
        declarations: [JitsiMeetAngularComponent],
        imports: [],
        exports: [JitsiMeetAngularComponent]
    })
], JitsiMeetAngularModule);

/*
 * Public API Surface of jitsi-meet-angular
 */

/**
 * Generated bundle index. Do not edit.
 */

export { JitsiMeetAngularComponent, JitsiMeetAngularModule, JitsiMeetAngularService };
