// @syncfusion/ej2-spreadsheet/src/spreadsheet/actions/show-hide.js
import { spreadsheetDestroyed, hideShowCol, hideShowRow } from '../common/index';
import { autoFit } from '../common/index';
import { getCellAddress, isHiddenRow, setRow } from '../../workbook/index';
import { detach } from '@syncfusion/ej2-base';
/**
 * The `ShowHide` module is used to perform hide/show the rows and columns.
 * @hidden
 */
var ShowHide = /** @class */ (function () {
    /**
     * Constructor for the Spreadsheet show hide module.
     * @private
     */
    function ShowHide(parent) {
        this.parent = parent;
        this.addEventListener();
    }
    // tslint:disable-next-line
    ShowHide.prototype.showHideRow = function (args) {
        var sheet = this.parent.getActiveSheet();
        var count = 0;
        var idx;
        if (args.hide) {
            var content = void 0;
            var rowHdr = void 0;
            var row = void 0;
            for (var i = args.startRow; i <= args.endRow; i++) {
                if (isHiddenRow(sheet, i)) {
                    continue;
                }
                if (idx === undefined) {
                    if (sheet.showHeaders) {
                        rowHdr = this.parent.getRowHeaderTable();
                    }
                    content = this.parent.getContentTable();
                    idx = this.getViewportIdx(i);
                    count = 0;
                }
                setRow(sheet, i, { hidden: true });
                row = content.rows[idx];
                if (row) {
                    if (sheet.showHeaders) {
                        detach(rowHdr.rows[idx]);
                    }
                    detach(row);
                    count++;
                }
                else {
                    count--;
                }
            }
            if (!count) {
                return;
            }
            this.parent.selectRange(sheet.selectedRange);
            if (this.parent.viewport.topIndex >= args.startRow) {
                this.parent.viewport.topIndex = args.endRow + 1;
            }
            if (this.parent.scrollSettings.enableVirtualization) {
                args.startRow = this.parent.viewport.bottomIndex + 1;
                args.endRow = args.startRow + count - 1;
                var indexes = this.parent.skipHiddenRows(args.startRow, args.endRow);
                args.startRow = indexes[0];
                args.endRow = indexes[1];
                this.parent.viewport.bottomIndex = args.endRow;
                this.parent.renderModule.refreshUI({ colIndex: this.parent.viewport.leftIndex, direction: '', refresh: 'RowPart' }, getCellAddress(args.startRow, this.parent.viewport.leftIndex) + ":" + getCellAddress(args.endRow, this.parent.viewport.leftIndex + this.parent.viewport.colCount + (this.parent.getThreshold('col') * 2)));
            }
            if (sheet.showHeaders) {
                if (idx === 0) {
                    rowHdr.rows[0].classList.add('e-hide-end');
                }
                else {
                    rowHdr.rows[idx - 1].classList.add('e-hide-start');
                    rowHdr.rows[idx].classList.add('e-hide-end');
                }
            }
        }
        else {
            var hFrag = void 0;
            var frag = void 0;
            var hRow = void 0;
            var rowRenderer = void 0;
            var rTBody = void 0;
            var tBody = void 0;
            var endRow = args.startRow - 1;
            var newStartRow = void 0;
            for (var i = args.startRow, len = args.endRow; i <= len; i++) {
                if (!isHiddenRow(sheet, i)) {
                    if (args.startRow === args.endRow) {
                        return;
                    }
                    if (idx === undefined) {
                        endRow++;
                    }
                    else {
                        newStartRow = i;
                    }
                    continue;
                }
                if (newStartRow !== undefined) {
                    len = i;
                    continue;
                }
                if (i > this.parent.viewport.bottomIndex) {
                    setRow(sheet, i, { hidden: false });
                    continue;
                }
                setRow(sheet, i, { hidden: false });
                if (idx === undefined) {
                    hFrag = document.createDocumentFragment();
                    frag = document.createDocumentFragment();
                    rowRenderer = this.parent.serviceLocator.getService('row');
                    if (sheet.showHeaders) {
                        rTBody = this.parent.getRowHeaderTable().tBodies[0];
                    }
                    tBody = this.parent.getContentTable().tBodies[0];
                    if (i < this.parent.viewport.topIndex) {
                        this.parent.viewport.topIndex = i;
                    }
                    if (i === 0) {
                        idx = 0;
                    }
                    else {
                        idx = this.getViewportIdx(i);
                    }
                }
                endRow++;
                hRow = rowRenderer.refresh(i);
                hFrag.appendChild(hRow);
                frag.appendChild(rowRenderer.refresh(i, hRow));
                if (sheet.showHeaders) {
                    detach(rTBody.lastElementChild);
                }
                detach(tBody.lastElementChild);
            }
            this.parent.viewport.bottomIndex = this.parent.viewport.topIndex + this.parent.viewport.rowCount +
                (this.parent.getThreshold('row') * 2);
            count = this.parent.hiddenRowsCount(this.parent.viewport.topIndex, args.startRow) +
                this.parent.hiddenRowsCount(args.endRow + 1, this.parent.viewport.bottomIndex);
            this.parent.viewport.bottomIndex += count;
            args.insertIdx = idx;
            args.row = frag.querySelector('.e-row');
            if (sheet.showHeaders) {
                args.hdrRow = hFrag.querySelector('.e-row');
                if (idx !== 0 && !isHiddenRow(sheet, endRow - 1)) {
                    rTBody.children[idx - 1].classList.remove('e-hide-start');
                }
                if (args.startRow !== 0 && isHiddenRow(sheet, args.startRow - 1)) {
                    args.hdrRow.classList.add('e-hide-end');
                }
                if (isHiddenRow(sheet, endRow + 1)) {
                    hFrag.lastElementChild.classList.add('e-hide-start');
                }
                else {
                    rTBody.children[idx].classList.remove('e-hide-end');
                }
            }
            if (args.skipAppend) {
                return;
            }
            if (sheet.showHeaders) {
                rTBody.insertBefore(hFrag, rTBody.children[idx]);
            }
            tBody.insertBefore(frag, tBody.children[idx]);
            this.parent.selectRange(sheet.selectedRange);
            if (args.autoFit && sheet.showHeaders) {
                this.parent.notify(autoFit, { startIndex: args.startRow, endIndex: args.endRow, isRow: true });
            }
            if (newStartRow !== undefined && newStartRow !== args.endRow) {
                args.startRow = newStartRow;
                this.showHideRow(args);
            }
        }
    };
    ShowHide.prototype.showHideCol = function () {
        /** */
    };
    ShowHide.prototype.getViewportIdx = function (index) {
        if (this.parent.scrollSettings.enableVirtualization) {
            index -= this.parent.hiddenRowsCount(this.parent.viewport.topIndex, index);
            index -= this.parent.viewport.topIndex;
        }
        return index;
    };
    ShowHide.prototype.addEventListener = function () {
        this.parent.on(hideShowRow, this.showHideRow, this);
        this.parent.on(hideShowCol, this.showHideCol, this);
        this.parent.on(spreadsheetDestroyed, this.destroy, this);
    };
    ShowHide.prototype.destroy = function () {
        this.removeEventListener();
        this.parent = null;
    };
    ShowHide.prototype.removeEventListener = function () {
        this.parent.off(hideShowRow, this.showHideRow);
        this.parent.off(hideShowCol, this.showHideCol);
        this.parent.off(spreadsheetDestroyed, this.destroy);
    };
    return ShowHide;
}());
export { ShowHide };
