// ng-chat/esm2015/ng-chat/core/chat-adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
export class ChatAdapter {
    constructor() {
        // ### Abstract adapter methods ###
        // Event handlers
        this.friendsListChangedHandler = (participantsResponse) => { };
        this.messageReceivedHandler = (participant, message) => { };
    }
    // ### Adapter/Chat income/ingress events ###
    /**
     * @param {?} participantsResponse
     * @return {?}
     */
    onFriendsListChanged(participantsResponse) {
        this.friendsListChangedHandler(participantsResponse);
    }
    /**
     * @param {?} participant
     * @param {?} message
     * @return {?}
     */
    onMessageReceived(participant, message) {
        this.messageReceivedHandler(participant, message);
    }
}
if (false) {
    /** @type {?} */
    ChatAdapter.prototype.friendsListChangedHandler;
    /** @type {?} */
    ChatAdapter.prototype.messageReceivedHandler;
    /**
     * @abstract
     * @return {?}
     */
    ChatAdapter.prototype.listFriends = function () { };
    /**
     * @abstract
     * @param {?} destinataryId
     * @return {?}
     */
    ChatAdapter.prototype.getMessageHistory = function (destinataryId) { };
    /**
     * @abstract
     * @param {?} message
     * @return {?}
     */
    ChatAdapter.prototype.sendMessage = function (message) { };
}
