// jitsi-meet-angular/fesm5/jitsi-meet-angular.js
import { __decorate } from 'tslib';
import { ɵɵdefineInjectable, Injectable, Component, NgModule } from '@angular/core';

var JitsiMeetAngularService = /** @class */ (function () {
    function JitsiMeetAngularService() {
    }
    JitsiMeetAngularService.ɵprov = ɵɵdefineInjectable({ factory: function JitsiMeetAngularService_Factory() { return new JitsiMeetAngularService(); }, token: JitsiMeetAngularService, providedIn: "root" });
    JitsiMeetAngularService = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], JitsiMeetAngularService);
    return JitsiMeetAngularService;
}());

var JitsiMeetAngularComponent = /** @class */ (function () {
    function JitsiMeetAngularComponent() {
    }
    JitsiMeetAngularComponent.prototype.ngOnInit = function () {
    };
    JitsiMeetAngularComponent = __decorate([
        Component({
            selector: 'lib-jitsi-meet-angular',
            template: "\n    <p>\n      hello from jitsi-meet-angular!\n    </p>\n  "
        })
    ], JitsiMeetAngularComponent);
    return JitsiMeetAngularComponent;
}());

var JitsiMeetAngularModule = /** @class */ (function () {
    function JitsiMeetAngularModule() {
    }
    JitsiMeetAngularModule = __decorate([
        NgModule({
            declarations: [JitsiMeetAngularComponent],
            imports: [],
            exports: [JitsiMeetAngularComponent]
        })
    ], JitsiMeetAngularModule);
    return JitsiMeetAngularModule;
}());

/*
 * Public API Surface of jitsi-meet-angular
 */

/**
 * Generated bundle index. Do not edit.
 */

export { JitsiMeetAngularComponent, JitsiMeetAngularModule, JitsiMeetAngularService };
