// ng-chat/esm5/ng-chat/core/chat-adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
ChatAdapter = /** @class */ (function () {
    function ChatAdapter() {
        // ### Abstract adapter methods ###
        // Event handlers
        this.friendsListChangedHandler = function (participantsResponse) { };
        this.messageReceivedHandler = function (participant, message) { };
    }
    // ### Adapter/Chat income/ingress events ###
    // ### Adapter/Chat income/ingress events ###
    /**
     * @param {?} participantsResponse
     * @return {?}
     */
    ChatAdapter.prototype.onFriendsListChanged = 
    // ### Adapter/Chat income/ingress events ###
    /**
     * @param {?} participantsResponse
     * @return {?}
     */
    function (participantsResponse) {
        this.friendsListChangedHandler(participantsResponse);
    };
    /**
     * @param {?} participant
     * @param {?} message
     * @return {?}
     */
    ChatAdapter.prototype.onMessageReceived = /**
     * @param {?} participant
     * @param {?} message
     * @return {?}
     */
    function (participant, message) {
        this.messageReceivedHandler(participant, message);
    };
    return ChatAdapter;
}());
/**
 * @abstract
 */
export { ChatAdapter };
if (false) {
    /** @type {?} */
    ChatAdapter.prototype.friendsListChangedHandler;
    /** @type {?} */
    ChatAdapter.prototype.messageReceivedHandler;
    /**
     * @abstract
     * @return {?}
     */
    ChatAdapter.prototype.listFriends = function () { };
    /**
     * @abstract
     * @param {?} destinataryId
     * @return {?}
     */
    ChatAdapter.prototype.getMessageHistory = function (destinataryId) { };
    /**
     * @abstract
     * @param {?} message
     * @return {?}
     */
    ChatAdapter.prototype.sendMessage = function (message) { };
}
