// ng-chat/esm5/ng-chat/core/group.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Guid } from "./guid";
import { ChatParticipantStatus } from "./chat-participant-status.enum";
import { ChatParticipantType } from "./chat-participant-type.enum";
var Group = /** @class */ (function () {
    function Group(participants) {
        this.id = Guid.newGuid();
        this.participantType = ChatParticipantType.Group;
        this.chattingTo = participants;
        this.status = ChatParticipantStatus.Online;
        // TODO: Add some customization for this in future releases
        this.displayName = participants.map(function (p) { return p.displayName; }).sort(function (first, second) { return second > first ? -1 : 1; }).join(", ");
    }
    return Group;
}());
export { Group };
if (false) {
    /** @type {?} */
    Group.prototype.id;
    /** @type {?} */
    Group.prototype.chattingTo;
    /** @type {?} */
    Group.prototype.participantType;
    /** @type {?} */
    Group.prototype.status;
    /** @type {?} */
    Group.prototype.avatar;
    /** @type {?} */
    Group.prototype.displayName;
}
