// ng-chat/esm5/ng-chat/core/window.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
var Window = /** @class */ (function () {
    function Window(participant, isLoadingHistory, isCollapsed) {
        this.messages = [];
        this.newMessage = "";
        // UI Behavior properties
        this.isCollapsed = false;
        this.isLoadingHistory = false;
        this.hasFocus = false;
        this.hasMoreMessages = true;
        this.historyPage = 0;
        this.participant = participant;
        this.messages = [];
        this.isLoadingHistory = isLoadingHistory;
        this.hasFocus = false; // This will be triggered when the 'newMessage' input gets the current focus
        this.isCollapsed = isCollapsed;
        this.hasMoreMessages = false;
        this.historyPage = 0;
    }
    return Window;
}());
export { Window };
if (false) {
    /** @type {?} */
    Window.prototype.participant;
    /** @type {?} */
    Window.prototype.messages;
    /** @type {?} */
    Window.prototype.newMessage;
    /** @type {?} */
    Window.prototype.isCollapsed;
    /** @type {?} */
    Window.prototype.isLoadingHistory;
    /** @type {?} */
    Window.prototype.hasFocus;
    /** @type {?} */
    Window.prototype.hasMoreMessages;
    /** @type {?} */
    Window.prototype.historyPage;
}
