// @angular-material-extensions/fab-menu/esm5/lib/mat-fab-menu.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { speedDialFabAnimations } from './mat-fab-menu.animations';
var MatFabMenuComponent = /** @class */ (function () {
    function MatFabMenuComponent() {
        this.icon = 'add';
        this.direction = 'top';
        this.color = 'accent';
        this.closeAfterSelection = true;
        // tslint:disable-next-line:no-output-on-prefix
        this.onFabMenuItemSelected = new EventEmitter();
    }
    MatFabMenuComponent.prototype.ngOnInit = function () {
        this.adjustLayout();
    };
    MatFabMenuComponent.prototype.ngOnChanges = function (changes) {
        if (changes.direction && !changes.direction.firstChange) {
            this.direction = changes.direction.currentValue;
            this.adjustLayout();
        }
        if (changes.color && !changes.color.firstChange) {
            this.color = changes.color.currentValue;
        }
        if (changes.fabButtons && !changes.fabButtons.firstChange) {
            this.fabButtons = changes.fabButtons.currentValue;
        }
    };
    MatFabMenuComponent.prototype.adjustLayout = function () {
        switch (this.direction) {
            case 'top':
                this.layout = 'column-reverse';
                this.layout2 = 'column-reverse';
                break;
            case 'bottom':
                this.layout = 'column';
                this.layout2 = 'column';
                break;
            case 'left':
                this.layout = 'row-reverse';
                this.layout2 = 'row-reverse';
                break;
            case 'right':
                this.layout = 'row';
                this.layout2 = 'row';
                break;
        }
    };
    MatFabMenuComponent.prototype.toggle = function () {
        this.isActive = !this.isActive;
    };
    MatFabMenuComponent.prototype.selectFabMenu = function (fab) {
        this.onFabMenuItemSelected.emit(fab.id);
        if (this.closeAfterSelection) {
            this.isActive = false;
        }
    };
    __decorate([
        Input()
    ], MatFabMenuComponent.prototype, "fabButtons", void 0);
    __decorate([
        Input()
    ], MatFabMenuComponent.prototype, "icon", void 0);
    __decorate([
        Input()
    ], MatFabMenuComponent.prototype, "direction", void 0);
    __decorate([
        Input()
    ], MatFabMenuComponent.prototype, "color", void 0);
    __decorate([
        Input()
    ], MatFabMenuComponent.prototype, "isActive", void 0);
    __decorate([
        Input()
    ], MatFabMenuComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], MatFabMenuComponent.prototype, "closeAfterSelection", void 0);
    __decorate([
        Output()
    ], MatFabMenuComponent.prototype, "onFabMenuItemSelected", void 0);
    MatFabMenuComponent = __decorate([
        Component({
            selector: 'mat-fab-menu',
            template: "<div [fxLayout]=\"layout\" class=\"container\" fxLayoutAlign=\"center center\" fxLayoutGap=\"16px\">\n  <div fxLayoutAlign=\"center\">\n    <button (click)=\"isActive = !isActive\"\n            [color]=\"color\" [disabled]=\"disabled\"\n            [ngClass]=\"{\n    'mat-fab' : !isActive,\n    'mat-mini-fab' : isActive\n  }\"\n            mat-fab>\n      <mat-icon [@fabToggler]=\"{value: isActive}\">{{icon}}</mat-icon> <!-- Animation here -->\n    </button>\n  </div>\n  <div *ngIf=\"isActive\" [@fabsStagger]=\"fabButtons.length\"\n       [fxLayout]=\"layout2\" fxLayoutAlign=\"center center\" fxLayoutGap=\"16px\">\n    <button (click)=\"selectFabMenu(fab)\"\n            *ngFor=\"let fab of fabButtons\"\n            [color]=\"fab?.color\"\n            [matTooltip]=\"fab?.tooltip\"\n            [matTooltipPosition]=\"fab?.tooltipPosition\"\n            [ngClass]=\"{'fullSVG' : fab?.imgUrl}\"\n            mat-fab>\n      <mat-icon *ngIf=\"fab?.icon\" [color]=\"fab?.iconColor\">{{fab?.icon}}</mat-icon>\n      <img *ngIf=\"fab?.imgUrl\" [src]=\"fab?.imgUrl\" alt=\"icon\">\n    </button>\n  </div>\n</div>\n",
            animations: speedDialFabAnimations,
            styles: [":host{display:block}.fullSVG img{width:100%}.fullSVG ::ng-deep span.mat-button-wrapper{padding:0}.container>div{margin-bottom:5px}.container>div button{margin-bottom:16px}"]
        })
    ], MatFabMenuComponent);
    return MatFabMenuComponent;
}());
export { MatFabMenuComponent };
